src/crypto/secretstream/secretstream_macros.rs

```diff
@@ -0,0 +1,353 @@
+                &mut state,
+                header.0.as_ptr(),
+                key.0.as_ptr(),
+            )
+        };
+        if rc == -1 {
+            // NOTE: this return code explicitly means the header is invalid,
+            // but when implementing error types we should still consider the
+            // possibility of some other non-zero code below with a generic call
+            // to external function failed error.
+            return Err(());
+        } else if rc != 0 {
+            return Err(());
+        }
+        Ok(Stream::<Pull> {
+            state: state,
+            finalized: false,
+            phantom: core::marker::PhantomData,
+        })
+    }
+
+    /// Verifies that `c` is a valid ciphertext with a correct authentication tag
+    /// given the internal state of the `Stream` (ciphertext streams cannot be
+    /// decrypted out of order for this reason). Also may validate the optional
+    /// unencrypted additional data `ad` using the authentication tag attached to
+    /// `c`. Finally decrypts the ciphertext and tag, and checks the tag
+    /// validity.
+    /// If any authentication fails, the stream has already been finalized, or if
+    /// the tag byte for some reason does not correspond to a valid `Tag`,
+    /// returns `Err(())`. Otherwise returns the plaintext and the tag.
+    /// Applications will typically use a `while stream.is_not_finalized()`
+    /// loop to authenticate and decrypt a stream of messages.
+    pub fn pull(&mut self, c: &[u8], ad: Option<&[u8]>) -> Result<(Vec<u8>, Tag), ()> {
+        if self.finalized {
+            return Err(());
+        }
+        let c_len = c.len();
+        if c_len < ABYTES {
+            // An empty message will still be at least ABYTES.
+            return Err(());
+        }
+        let m_len = c_len - ABYTES;
+        if m_len > messagebytes_max() {
+            return Err(());
+        }
+        let (ad_p, ad_len) = ad
+            .map(|ad| (ad.as_ptr(), ad.len()))
+            .unwrap_or((ptr::null(), 0));
+        let mut tag: u8 = unsafe { mem::uninitialized() };
+        let mut buf = Vec::with_capacity(m_len);
+        let rc = unsafe {
+            buf.set_len(m_len);
+            $pull_name(
+                &mut self.state,
+                buf.as_mut_ptr(),
+                &mut (m_len as c_ulonglong),
+                &mut tag,
+                c.as_ptr(),
+                c_len as c_ulonglong,
+                ad_p,
+                ad_len as c_ulonglong,
+            )
+        };
+        if rc != 0 {
+            return Err(());
+        }
+        let tag = Tag::from_u8(tag)?;
+        if tag == Tag::Final {
+            self.finalized = true;
+        }
+        Ok((buf, tag))
+    }
+}
+
+// As additional precaution, rotate the keys when dropping the `Stream`
+// to ensure keys do no stay in memory.
+impl<T: StreamMode> Drop for Stream<T> {
+    fn drop(&mut self) {
+        let _ = self.rekey();
```

Comment:
Does `new_type! secret Key(KEYBYTES);` add impl Drop for Key with drop=sodium_memzero?

Reply:
Yes (see [here](https://github.com/sodiumoxide/sodiumoxide/blob/7512559dc5c5404dc8e3edbda76e1297a393dc11/src/newtype_macros.rs#L213)).

Reply:
So there is no sense in `let _ = self.rekey();` with comment `// to ensure keys do no stay in memory.`

Reply:
I guess the reason why this is here is the make sure the current internal key (stored in `crypto_secretstream_xchacha20poly1305_state`) is replaced by a 'useless' key. The new type `Key` is not used here (`rekey` returns `Result<(), ()>`).

Reply:
Ah I see, I've now removed the redundant rekeying in 2f0500e2c45e26122ba2f01153a64968073a9a89.